contract/src/discount/state.rs

```diff
@@ -0,0 +1,460 @@
+                    remain_available_for_sale.saturating_sub(sale_tokens_per_deposit);
+                sale_tokens_for_prev_phases =
+                    sale_tokens_for_prev_phases.saturating_add(sale_tokens_per_deposit);
+            }
+
+            // No more deposit or nothing to sell.
+            if remain_deposit == 0 || remain_available_for_sale == 0 {
+                break;
+            }
+        }
+
+        // There are available tokens for sale and remain deposit tokens, which we can spend them for public sale.
+        let public_sale_weight =
+            if remain_deposit > 0 && remain_available_for_sale > 0 && is_public_sale_allowed {
+                let sale_tokens =
+                    calculate_amount_of_sale_tokens(remain_deposit, deposit_token.0, sale_token.0)?;
+                let exceeded_global_limit = sale_tokens.saturating_sub(remain_available_for_sale);
+
+                if exceeded_global_limit > 0 {
+                    let exceeded_deposit = calculate_weight_from_sale_tokens(
+                        exceeded_global_limit,
+                        deposit_token.0,
+                        sale_token.0,
+                    )?;
+
+                    refund = exceeded_deposit;
+                    remain_deposit.saturating_sub(exceeded_deposit)
+                } else {
+                    remain_deposit
+                }
+            } else {
+                refund = remain_deposit;
+                0
+            };
+
+        Ok(DepositDistribution::WithDiscount {
+            phase_weights,
+            public_sale_weight,
+            refund,
+        })
+    }
+
+    fn get_account_sale_tokens_for_phase(&self, account: &IntentsAccount, phase_id: u16) -> u128 {
+        self.phases
+            .get(&phase_id)
+            .and_then(|phase_state| phase_state.account_sale_tokens.get(account))
+            .copied()
+            .unwrap_or(0)
+    }
+
+    fn get_total_sale_tokens_for_phases_with_limits(&self, phase_id: u16) -> u128 {
+        self.phases
+            .iter()
+            .filter(|(id, phase_state)| **id <= phase_id && phase_state.limit_per_phase.is_some())
+            .filter(|(id, _)| **id == phase_id || self.is_phases_linked(phase_id, **id))
+            .map(|(_, phase_state)| phase_state.total_sale_tokens)
+            .sum()
+    }
+
+    fn check_sale_phases_limit_exceeded(&self, sale_tokens: u128, phase_id: u16) -> u128 {
+        let current_phase_limit = self
+            .phases
+            .get(&phase_id)
+            .and_then(|phase_state| phase_state.limit_per_phase);
+
+        // We don't care if the current phase has no limit. Spend full deposit for sale tokens.
+        if current_phase_limit.is_none() {
+            return 0;
+        }
+
+        // ID of phases that share their limits with the current phase.
+        let total_limits = self
+            .phases
+            .iter()
+            .filter(|(id, _)| **id <= phase_id)
```

Comment:
I think this `filter` can be deleted because it is made redundant by the stricter condition below it.